numpy/core/tests/test_multiarray.py

```diff
@@ -4811,17 +4810,27 @@ def test_invalid_axis(self): # gh-7528
 class TestIO:
     """Test tofile, fromfile, tobytes, and fromstring"""
 
+    @pytest.fixture(autouse=True)
     def setup(self):
         shape = (2, 4, 3)
         rand = np.random.random
         self.x = rand(shape) + rand(shape).astype(complex)*1j
         self.x[0,:, 1] = [np.nan, np.inf, -np.inf, np.nan]
         self.dtype = self.x.dtype
-        self.tempdir = tempfile.mkdtemp()
-        self.filename = tempfile.mktemp(dir=self.tempdir)
-
-    def teardown(self):
-        shutil.rmtree(self.tempdir)
+        yield
+        delattr(self, "dtype")
+        delattr(self, "x")
```

Comment:
I think `setup/teardown` is a better paradigm here than using a `yield` like this. Why do you want to re-create `dtype` and `x`?

Reply:
In this case then I think it would be clearer if `x` was passed in to the tests that need it as an additional argument, rather than stuffing it onto `self`. Tests that need `dtype` should just call `x.dtype`. As I commented elsewhere, this use/abuse of `self.x` seems weird.

Reply:
I refactored `TestIO` to use a fixture (`x`) and also updated all the `self.filename` references to use the value from the `tmp_filename` fixture instead, which remove all attributes stored in `self`.

Is it too much out of scope to refactor `TestMemmap` to use fixtures too since the original issue was to remove the deprecated `mktemp()` function?

Reply:
I think the `TestMemmap` is fine as is. I would like to keep the use of fixtures to a minimum, it adds a layer of indirection that is harder to parse.